File: Projects/Offensive/Steganography encoder/steganography.py
#!/usr/bin/env python3
"""
Steganography Tool - Hide secret data in images and audio files
Purpose: Covert communication, data protection, watermarking
Use: Hide sensitive information in carrier files
"""

from PIL import Image
import wave
import argparse
import os

class ImageSteganography:
    def __init__(self):
        self.STRING_TERMINATOR = "###END###"
    
    def text_to_binary(self, text):
        """Convert text to binary string"""
        return ''.join(format(ord(char), '08b') for char in text)
    
    def binary_to_text(self, binary_str):
        """Convert binary string to text"""
        text = ''
        for i in range(0, len(binary_str), 8):
            byte = binary_str[i:i+8]
            text += chr(int(byte, 2))
        return text
    
    def encode_image(self, image_path, secret_data, output_path):
        """
        Encode secret data into image using LSB (Least Significant Bit) technique
        """
        try:
            # Open image and convert to RGB
            image = Image.open(image_path)
            pixels = image.load()
            width, height = image.size
            
            # Convert secret data to binary
            binary_data = self.text_to_binary(secret_data + self.STRING_TERMINATOR)
            data_len = len(binary_data)
            
            # Check if image can hold the data
            if data_len > width * height * 3:
                raise ValueError("Image too small to hold the secret data")
            
            data_index = 0
            
            # Encode data in pixels
            for y in range(height):
                for x in range(width):
                    if data_index < data_len:
                        pixel = list(pixels[x, y])
                        
                        # Modify RGB values
                        for color in range(3):
                            if data_index < data_len:
                                # Clear LSB and set to data bit
                                pixel[color] = (pixel[color] & 0xFE) | int(binary_data[data_index])
                                data_index += 1
                        
                        pixels[x, y] = tuple(pixel)
                    else:
                        break
                else:
                    continue
                break
            
            # Save encoded image
            image.save(output_path)
            print(f"Data encoded successfully in {output_path}")
            return True
            
        except Exception as e:
            print(f"Encoding failed: {str(e)}")
            return False
    
    def decode_image(self, image_path):
        """
        Decode secret data from image
        """
        try:
            image = Image.open(image_path)
            pixels = image.load()
            width, height = image.size
            
            binary_data = ""
            
            # Extract LSB from each pixel
            for y in range(height):
                for x in range(width):
                    pixel = pixels[x, y]
                    for color in range(3):
                        # Get LSB from each color channel
                        binary_data += str(pixel[color] & 1)
            
            # Convert binary to text
            text = self.binary_to_text(binary_data)
            
            # Find terminator
            if self.STRING_TERMINATOR in text:
                secret_data = text.split(self.STRING_TERMINATOR)[0]
                print(f"Decoded data: {secret_data}")
                return secret_data
            else:
                print("No hidden data found or data corrupted")
                return None
                
        except Exception as e:
            print(f"Decoding failed: {str(e)}")
            return None

class AudioSteganography:
    def __init__(self):
        self.STRING_TERMINATOR = "###END###"
    
    def text_to_binary(self, text):
        """Convert text to binary string"""
        return ''.join(format(ord(char), '08b') for char in text)
    
    def binary_to_text(self, binary_str):
        """Convert binary string to text"""
        text = ''
        for i in range(0, len(binary_str), 8):
            byte = binary_str[i:i+8]
            text += chr(int(byte, 2))
        return text
    
    def encode_audio(self, audio_path, secret_data, output_path):
        """
        Encode secret data into audio file using LSB technique
        """
        try:
            audio = wave.open(audio_path, 'rb')
            params = audio.getparams()
            frames = audio.readframes(audio.getnframes())
            audio.close()
            
            # Convert frames to byte array
            frame_bytes = bytearray(list(frames))
            
            # Convert secret data to binary
            binary_data = self.text_to_binary(secret_data + self.STRING_TERMINATOR)
            data_len = len(binary_data)
            
            # Check if audio can hold the data
            if data_len > len(frame_bytes):
                raise ValueError("Audio file too small to hold the secret data")
            
            # Encode data in audio frames
            for i in range(data_len):
                frame_bytes[i] = (frame_bytes[i] & 0xFE) | int(binary_data[i])
            
            # Save encoded audio
            with wave.open(output_path, 'wb') as encoded_audio:
                encoded_audio.setparams(params)
                encoded_audio.writeframes(bytes(frame_bytes))
            
            print(f"Data encoded successfully in {output_path}")
            return True
            
        except Exception as e:
            print(f"Audio encoding failed: {str(e)}")
            return False
    
    def decode_audio(self, audio_path):
        """
        Decode secret data from audio file
        """
        try:
            audio = wave.open(audio_path, 'rb')
            frames = audio.readframes(audio.getnframes())
            audio.close()
            
            # Extract LSB from each frame
            frame_bytes = bytearray(list(frames))
            binary_data = ''.join(str(byte & 1) for byte in frame_bytes)
            
            # Convert binary to text
            text = self.binary_to_text(binary_data)
            
            # Find terminator
            if self.STRING_TERMINATOR in text:
                secret_data = text.split(self.STRING_TERMINATOR)[0]
                print(f"Decoded data: {secret_data}")
                return secret_data
            else:
                print("No hidden data found or data corrupted")
                return None
                
        except Exception as e:
            print(f"Audio decoding failed: {str(e)}")
            return None

def main():
    parser = argparse.ArgumentParser(description='Steganography Tool - Hide data in images and audio')
    parser.add_argument('--mode', choices=['encode', 'decode'], required=True, help='Operation mode')
    parser.add_argument('--type', choices=['image', 'audio'], required=True, help='Carrier file type')
    parser.add_argument('--input', required=True, help='Input file path')
    parser.add_argument('--output', help='Output file path (for encoding)')
    parser.add_argument('--data', help='Secret data to hide (for encoding)')
    
    args = parser.parse_args()
    
    if args.type == 'image':
        stego = ImageSteganography()
    else:
        stego = AudioSteganography()
    
    if args.mode == 'encode':
        if not args.data or not args.output:
            print("Error: --data and --output required for encoding")
            return
        stego.encode_image(args.input, args.data, args.output) if args.type == 'image' else stego.encode_audio(args.input, args.data, args.output)
    else:
        stego.decode_image(args.input) if args.type == 'image' else stego.decode_audio(args.input)

if __name__ == "__main__":
    main()

File: Projects/Offensive/Steganography encoder/steganography.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <math.h>

#define STRING_TERMINATOR "###END###"
#define MAX_DATA_SIZE 10000

typedef struct {
    unsigned char red;
    unsigned char green;
    unsigned char blue;
} Pixel;

typedef struct {
    char type[3];
    int width;
    int height;
    int max_value;
    Pixel **pixels;
} PPMImage;

// Utility functions
void text_to_binary(const char *text, char *binary) {
    int i, j;
    char *ptr = binary;
    
    for (i = 0; text[i] != '\0'; i++) {
        unsigned char ch = text[i];
        for (j = 7; j >= 0; j--) {
            *ptr++ = ((ch >> j) & 1) ? '1' : '0';
        }
    }
    *ptr = '\0';
}

void binary_to_text(const char *binary, char *text) {
    int i, j = 0;
    char byte[9];
    
    for (i = 0; binary[i] != '\0'; i += 8) {
        if (i + 8 <= strlen(binary)) {
            strncpy(byte, &binary[i], 8);
            byte[8] = '\0';
            text[j++] = (char)strtol(byte, NULL, 2);
        }
    }
    text[j] = '\0';
}

// PPM Image functions
PPMImage* read_ppm(const char *filename) {
    FILE *file = fopen(filename, "rb");
    if (!file) {
        printf("Error: Cannot open file %s\n", filename);
        return NULL;
    }
    
    PPMImage *image = (PPMImage*)malloc(sizeof(PPMImage));
    fscanf(file, "%2s", image->type);
    fscanf(file, "%d %d", &image->width, &image->height);
    fscanf(file, "%d", &image->max_value);
    fgetc(file); // Skip newline
    
    // Allocate memory for pixels
    image->pixels = (Pixel**)malloc(image->height * sizeof(Pixel*));
    for (int i = 0; i < image->height; i++) {
        image->pixels[i] = (Pixel*)malloc(image->width * sizeof(Pixel));
    }
    
    // Read pixel data
    for (int i = 0; i < image->height; i++) {
        for (int j = 0; j < image->width; j++) {
            image->pixels[i][j].red = fgetc(file);
            image->pixels[i][j].green = fgetc(file);
            image->pixels[i][j].blue = fgetc(file);
        }
    }
    
    fclose(file);
    return image;
}

int write_ppm(const char *filename, PPMImage *image) {
    FILE *file = fopen(filename, "wb");
    if (!file) {
        printf("Error: Cannot create file %s\n", filename);
        return 0;
    }
    
    fprintf(file, "%s\n", image->type);
    fprintf(file, "%d %d\n", image->width, image->height);
    fprintf(file, "%d\n", image->max_value);
    
    for (int i = 0; i < image->height; i++) {
        for (int j = 0; j < image->width; j++) {
            fputc(image->pixels[i][j].red, file);
            fputc(image->pixels[i][j].green, file);
            fputc(image->pixels[i][j].blue, file);
        }
    }
    
    fclose(file);
    return 1;
}

void free_ppm(PPMImage *image) {
    for (int i = 0; i < image->height; i++) {
        free(image->pixels[i]);
    }
    free(image->pixels);
    free(image);
}

// Steganography functions
int encode_ppm(PPMImage *image, const char *secret_data) {
    char binary_data[MAX_DATA_SIZE];
    char full_data[1024];
    
    // Prepare data with terminator
    snprintf(full_data, sizeof(full_data), "%s%s", secret_data, STRING_TERMINATOR);
    text_to_binary(full_data, binary_data);
    
    int data_len = strlen(binary_data);
    int max_capacity = image->width * image->height * 3;
    
    if (data_len > max_capacity) {
        printf("Error: Image too small to hold data\n");
        return 0;
    }
    
    int data_index = 0;
    
    // Encode data using LSB
    for (int i = 0; i < image->height && data_index < data_len; i++) {
        for (int j = 0; j < image->width && data_index < data_len; j++) {
            // Encode in red channel
            if (data_index < data_len) {
                image->pixels[i][j].red = (image->pixels[i][j].red & 0xFE) | (binary_data[data_index] - '0');
                data_index++;
            }
            
            // Encode in green channel
            if (data_index < data_len) {
                image->pixels[i][j].green = (image->pixels[i][j].green & 0xFE) | (binary_data[data_index] - '0');
                data_index++;
            }
            
            // Encode in blue channel
            if (data_index < data_len) {
                image->pixels[i][j].blue = (image->pixels[i][j].blue & 0xFE) | (binary_data[data_index] - '0');
                data_index++;
            }
        }
    }
    
    printf("Data encoded successfully\n");
    return 1;
}

char* decode_ppm(PPMImage *image) {
    int max_bits = image->width * image->height * 3;
    char *binary_data = (char*)malloc(max_bits + 1);
    int bit_index = 0;
    
    // Extract LSB from each pixel
    for (int i = 0; i < image->height; i++) {
        for (int j = 0; j < image->width; j++) {
            binary_data[bit_index++] = (image->pixels[i][j].red & 1) ? '1' : '0';
            binary_data[bit_index++] = (image->pixels[i][j].green & 1) ? '1' : '0';
            binary_data[bit_index++] = (image->pixels[i][j].blue & 1) ? '1' : '0';
        }
    }
    binary_data[bit_index] = '\0';
    
    // Convert to text
    char *text = (char*)malloc(bit_index / 8 + 1);
    binary_to_text(binary_data, text);
    
    free(binary_data);
    
    // Check for terminator
    char *terminator_pos = strstr(text, STRING_TERMINATOR);
    if (terminator_pos) {
        *terminator_pos = '\0';
        return text;
    }
    
    free(text);
    return NULL;
}

int main(int argc, char *argv[]) {
    if (argc < 4) {
        printf("Usage:\n");
        printf("  Encode: %s encode input.ppm output.ppm \"secret data\"\n", argv[0]);
        printf("  Decode: %s decode input.ppm\n", argv[0]);
        return 1;
    }
    
    if (strcmp(argv[1], "encode") == 0) {
        PPMImage *image = read_ppm(argv[2]);
        if (!image) return 1;
        
        if (encode_ppm(image, argv[4])) {
            write_ppm(argv[3], image);
        }
        
        free_ppm(image);
    }
    else if (strcmp(argv[1], "decode") == 0) {
        PPMImage *image = read_ppm(argv[2]);
        if (!image) return 1;
        
        char *decoded = decode_ppm(image);
        if (decoded) {
            printf("Decoded data: %s\n", decoded);
            free(decoded);
        } else {
            printf("No hidden data found\n");
        }
        
        free_ppm(image);
    }
    
    return 0;
}
